src/workerd/api/urlpattern.c++

```diff
@@ -3,1822 +3,47 @@
-}
-
-URLPatternComponents init(jsg::Lock& js, URLPattern::URLPatternInit&& init) {
-  KJ_IF_SOME(protocol, init.protocol) {
-    KJ_IF_SOME(port, init.port) {
-      if (url::URL::isSpecialScheme(protocol) &&
-          isSpecialSchemeDefaultPort(protocol, port)) {
-        init.port = jsg::usv();
-      }
-    }
-  }
-
-  auto protocolComponent = compileComponent(
-      js, init.protocol.map([](jsg::UsvString &str) { return str.asPtr(); }),
-      &canonicalizeProtocol, CompileOptions::DEFAULT);
-
-  auto matchesSpecialScheme = protocolComponentMatchesSpecialScheme(js, protocolComponent);
-
-  return URLPatternComponents{
-      .protocol = kj::mv(protocolComponent),
-      .username = compileComponent(
-          js, init.username.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          &canonicalizeUsername, CompileOptions::DEFAULT),
-      .password = compileComponent(
-          js, init.password.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          &canonicalizePassword, CompileOptions::DEFAULT),
-      .hostname = compileHostnameComponent(
-          js, init.hostname.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          CompileOptions::DEFAULT),
-      .port = compileComponent(
-          js, init.port.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          &canonicalizePort, CompileOptions::DEFAULT),
-      .pathname = compileComponent(
-          js, init.pathname.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          matchesSpecialScheme ? &canonicalizePathname
-                               : &canonicalizeOpaquePathname,
-          matchesSpecialScheme ? CompileOptions::PATHNAME : CompileOptions::DEFAULT),
-      .search = compileComponent(
-          js, init.search.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          &canonicalizeSearch, CompileOptions::DEFAULT),
-      .hash = compileComponent(
-          js, init.hash.map([](jsg::UsvString &str) { return str.asPtr(); }),
-          &canonicalizeHash, CompileOptions::DEFAULT),
-  };
-}
-
-URLPatternComponents init(
-    jsg::Lock& js,
-    jsg::Optional<URLPattern::URLPatternInput> maybeInput,
-    jsg::Optional<jsg::UsvString> baseURL) {
-  auto input = kj::mv(maybeInput).orDefault(URLPattern::URLPatternInit());
-  KJ_SWITCH_ONEOF(input) {
-    KJ_CASE_ONEOF(string, jsg::UsvString) {
-      return init(js, parseConstructorString(js, string, kj::mv(baseURL)));
-    }
-    KJ_CASE_ONEOF(i, URLPattern::URLPatternInit) {
-      JSG_REQUIRE(baseURL == nullptr, TypeError, BASEURL_ERROR);
-      return init(js, processPatternInit(i, ProcessPatternInitType::PATTERN));
-    }
-  }
-  KJ_UNREACHABLE;
+jsg::Ref<URLPattern> create(jsg::Lock& js, jsg::UrlPattern pattern) {
+  bool ignoreCase = pattern.getIgnoreCase();
+  return jsg::alloc<URLPattern>(kj::mv(pattern),
+      compileRegex(js, pattern.getProtocol(), ignoreCase),
+      compileRegex(js, pattern.getUsername(), ignoreCase),
```

Comment:
This is a bit boggling to the eye. `kj::mv` must happen after all the accessor calls that follow (not a complaint, but jarring to observe with the accessors of `pattern` following due to the execution order).

The forwarded arguments go to macro-ized `UrlPattern` constructor that takes _N_-regex arguments. It would be good to use a corresponding macro pattern, using URL_PATTERN_COMPONENTS, so it is trivially obvious the regexes line up.

Reply:
heh, that's fair. Reworking!